File: tasks/TimeSeries-Downsampler/steps/milestone_4/tests/test_m4.py
"""Tests for Milestone 4: Cascading Eviction."""
import os
import shutil
import subprocess
import tempfile

_MILESTONE = "m4"


def _run_rust_verifier():
    """Run the milestone's Rust assertions from a verifier-owned cargo harness.

    The harness is created in a temporary directory outside /app and depends on
    the agent's crate via a path dependency, so no verifier files are ever
    written into the agent's workspace. Returns the set of passing test names.
    """
    test_rs = os.path.join(
        os.path.dirname(os.path.abspath(__file__)), f"test_{_MILESTONE}.rs"
    )
    harness = tempfile.mkdtemp(prefix=f"verifier_{_MILESTONE}_")
    try:
        os.makedirs(os.path.join(harness, "src"), exist_ok=True)
        os.makedirs(os.path.join(harness, "tests"), exist_ok=True)
        with open(os.path.join(harness, "Cargo.toml"), "w") as fh:
            fh.write(
                "[package]\n"
                f'name = "verifier_{_MILESTONE}"\n'
                'version = "0.0.0"\n'
                'edition = "2021"\n\n'
                "[lib]\n"
                'path = "src/lib.rs"\n\n'
                "[dependencies]\n"
                'timeseries_downsampler = { path = "/app/app" }\n\n'
                "[[test]]\n"
                f'name = "test_{_MILESTONE}"\n'
                f'path = "tests/test_{_MILESTONE}.rs"\n'
            )
        open(os.path.join(harness, "src", "lib.rs"), "w").close()
        shutil.copyfile(
            test_rs, os.path.join(harness, "tests", f"test_{_MILESTONE}.rs")
        )
        result = subprocess.run(
            ["cargo", "test",
             "--manifest-path", os.path.join(harness, "Cargo.toml"),
             "--test", f"test_{_MILESTONE}", "--offline",
             "--", "--test-threads=1"],
            capture_output=True, text=True, timeout=300,
        )
        output = result.stdout + result.stderr
        passed = set()
        for line in output.splitlines():
            line = line.strip()
            parts = line.split(" ... ")
            if line.startswith("test ") and len(parts) == 2 and parts[1].strip() == "ok":
                passed.add(parts[0][5:])
        return passed
    except Exception:
        return set()
    finally:
        shutil.rmtree(harness, ignore_errors=True)


_PASSED = _run_rust_verifier()

class TestMilestone4:
    def test_evict_lowest_subtree_value(self):
        """Evicts the orphan with the lowest subtree value when over capacity."""
        assert "evict_lowest_subtree_value" in _PASSED

    def test_cascading_eviction_and_recalculation(self):
        """Evicting an event cascades, and capacity checking recalculates remaining subtree values."""
        assert "cascading_eviction_and_recalculation" in _PASSED

    def test_tie_breaking_by_event_id(self):
        """Ties in subtree value are broken by lexicographically smallest event_id."""
        assert "tie_breaking_by_event_id" in _PASSED

    def test_cascading_eviction_of_dependent_subtree(self):
        """Evicting an orphan must recursively evict its entire dependent subtree and clean up orphans map."""
        assert "cascading_eviction_of_dependent_subtree" in _PASSED

    def test_transitive_subtree_value_calculation(self):
        """Transitive dependencies must be correctly summed when calculating subtree values."""
        assert "transitive_subtree_value_calculation" in _PASSED

    def test_deduplication_in_return_value(self):
        """Evicted events returned should have exactly one instance even if they wait on multiple dependencies."""
        assert "deduplication_in_return_value" in _PASSED

File: tasks/TimeSeries-Downsampler/steps/milestone_4/tests/test_m4.rs
use timeseries_downsampler::causal::CausalBuffer;
use timeseries_downsampler::models::Event;

#[test]
fn evict_lowest_subtree_value() {
    let mut buf = CausalBuffer::new();
    // A depends on X. Value = 10
    buf.process_event(Event { timestamp_ms: 100, event_id: "A".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: vec!["X".to_string()] });
    // B depends on X. Value = 20
    buf.process_event(Event { timestamp_ms: 100, event_id: "B".to_string(), name: "cpu".to_string(), value: 20.0, dependency_ids: vec!["X".to_string()] });
    // C depends on X. Value = 5
    buf.process_event(Event { timestamp_ms: 100, event_id: "C".to_string(), name: "cpu".to_string(), value: 5.0, dependency_ids: vec!["X".to_string()] });
    
    // Subtree values: A=10, B=20, C=5. Capacity = 2. C should be evicted.
    let evicted = buf.evict_over_capacity(2);
    assert_eq!(evicted.len(), 1);
    assert_eq!(evicted[0].event_id, "C");
    assert!(buf.deadletter_ids.contains("C"));
}

#[test]
fn cascading_eviction_and_recalculation() {
    let mut buf = CausalBuffer::new();
    // A depends on X. Value = 10
    buf.process_event(Event { timestamp_ms: 100, event_id: "A".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: vec!["X".to_string()] });
    // B depends on A. Value = 50. Total Subtree(A) = 60
    buf.process_event(Event { timestamp_ms: 100, event_id: "B".to_string(), name: "cpu".to_string(), value: 50.0, dependency_ids: vec!["A".to_string()] });
    
    // C depends on X. Value = 40. Total Subtree(C) = 40
    buf.process_event(Event { timestamp_ms: 100, event_id: "C".to_string(), name: "cpu".to_string(), value: 40.0, dependency_ids: vec!["X".to_string()] });
    // D depends on C. Value = 10. Total Subtree(C) = 50. Total Subtree(D) = 10
    buf.process_event(Event { timestamp_ms: 100, event_id: "D".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: vec!["C".to_string()] });
    
    // If capacity = 2:
    // Pass 1: Lowest is D (10). Evict D. Remaining: A(60), B(50), C(40). Count = 3 > 2.
    // Pass 2: Lowest is C (40). Evict C. Remaining: A(60), B(50). Count = 2. Done.
    
    let mut evicted = buf.evict_over_capacity(2);
    evicted.sort_by(|a, b| a.event_id.cmp(&b.event_id));
    
    assert_eq!(evicted.len(), 2);
    assert_eq!(evicted[0].event_id, "C");
    assert_eq!(evicted[1].event_id, "D");
    assert!(buf.deadletter_ids.contains("C"));
    assert!(buf.deadletter_ids.contains("D"));
}

#[test]
fn tie_breaking_by_event_id() {
    let mut buf = CausalBuffer::new();
    // A depends on X. Value = 10
    buf.process_event(Event { timestamp_ms: 100, event_id: "Z".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: vec!["X".to_string()] });
    // B depends on X. Value = 10
    buf.process_event(Event { timestamp_ms: 100, event_id: "Y".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: vec!["X".to_string()] });
    
    // Tie between Z and Y. Lexicographically, Y < Z. So Y should be evicted if capacity = 1.
    let evicted = buf.evict_over_capacity(1);
    assert_eq!(evicted.len(), 1);
    assert_eq!(evicted[0].event_id, "Y");
}

#[test]
fn cascading_eviction_of_dependent_subtree() {
    let mut buf = CausalBuffer::new();
    buf.process_event(Event { timestamp_ms: 100, event_id: "A".to_string(), name: "cpu".to_string(), value: 100.0, dependency_ids: vec!["X".to_string()] });
    buf.process_event(Event { timestamp_ms: 100, event_id: "B".to_string(), name: "cpu".to_string(), value: 5.0, dependency_ids: vec!["X".to_string()] });
    buf.process_event(Event { timestamp_ms: 100, event_id: "C".to_string(), name: "cpu".to_string(), value: 1.0, dependency_ids: vec!["B".to_string()] });
    
    let evicted = buf.evict_over_capacity(1);
    assert_eq!(evicted.len(), 2);
    use std::collections::HashSet;
    let ids: HashSet<_> = evicted.iter().map(|e| e.event_id.as_str()).collect();
    assert!(ids.contains("B") && ids.contains("C"));
    assert!(!buf.deadletter_ids.contains("A"));

    // After eviction, verify no empty Vecs remain in orphans map
    for (_, v) in buf.orphans.iter() {
        assert!(!v.is_empty(), "Empty Vec left in orphans map");
    }
}

#[test]
fn transitive_subtree_value_calculation() {
    let mut buf = CausalBuffer::new();
    buf.process_event(Event { timestamp_ms: 100, event_id: "B".to_string(), name: "cpu".to_string(), value: 2.0, dependency_ids: vec!["A_dep".to_string()] });
    buf.process_event(Event { timestamp_ms: 100, event_id: "C".to_string(), name: "cpu".to_string(), value: 1.0, dependency_ids: vec!["B".to_string()] });
    buf.process_event(Event { timestamp_ms: 100, event_id: "D".to_string(), name: "cpu".to_string(), value: 10.0, dependency_ids: vec!["A_dep".to_string()] });
    
    let evicted = buf.evict_over_capacity(2);
    assert_eq!(evicted.len(), 1);
    assert_eq!(evicted[0].event_id, "C");
}

#[test]
fn deduplication_in_return_value() {
    let mut buf = CausalBuffer::new();
    buf.process_event(Event { timestamp_ms: 100, event_id: "E".to_string(), name: "cpu".to_string(), value: 1.0, dependency_ids: vec!["X".to_string(), "Y".to_string()] });
    let evicted = buf.evict_over_capacity(0);
    assert_eq!(evicted.iter().filter(|e| e.event_id == "E").count(), 1);
}
